File: Form1.vb
Public Class Form1
    Private i As Integer
    Private j As Integer
    Private turn As Integer
    Dim blackpoint As New List(Of Integer)
    Dim whitepoint As New List(Of Integer)
    Dim blacksum As Integer
    Dim whitesum As Integer
    Private Function colorchange(ByVal j As Integer) As Integer

        If (Button1.Enabled = False And Button2.Enabled = False And Button3.Enabled = False And Button4.Enabled = False And
                                                               Button1.BackColor = Color.White And Button4.BackColor = Color.White And i = 4) Then
            Button2.BackColor = Color.White
            Button3.BackColor = Color.White
            whitepoint.Add(2)
            blackpoint.Add(-2)
        End If

        ' If (Button1.Enabled = False And Button2.Enabled = False And Button3.Enabled = False And
        'Button1.BackColor = Color.White And Button3.BackColor = Color.White And i = 3) Then
        'Button2.BackColor = Color.White
        'End If

        If (Button2.Enabled = False And Button3.Enabled = False And Button4.Enabled = False And
                       Button2.BackColor = Color.White And Button4.BackColor = Color.White And i = 3) Then
            Button3.BackColor = Color.White
            whitepoint.Add(1)
            blackpoint.Add(-1)
        End If

        ' If (Button1.Enabled = False And Button2.Enabled = False And Button3.Enabled = False And Button4.Enabled = False And
        'Button1.BackColor = Color.Black And Button4.BackColor = Color.Black) Then
        'Button2.BackColor = Color.Black
        'Button3.BackColor = Color.Black
        'End If

        If (Button1.Enabled = False And Button2.Enabled = False And Button3.Enabled = False And
                 Button1.BackColor = Color.Black And Button3.BackColor = Color.Black And i = 3) Then
            Button2.BackColor = Color.Black
            blackpoint.Add(1)
            whitepoint.Add(-1)
        End If

        If (Button2.Enabled = False And Button3.Enabled = False And Button4.Enabled = False And
                       Button2.BackColor = Color.Black And Button4.BackColor = Color.Black And i = 3) Then
            Button3.BackColor = Color.Black
            blackpoint.Add(1)
            whitepoint.Add(-1)
        End If

        Return j

    End Function

    Private Sub button_click(ByVal sender As Object, e As EventArgs) Handles Button1.Click, Button2.Click, Button3.Click, Button4.Click
        i = i + 1
        turn = i Mod 2

        Label1.Text = Int(i)
        If turn = 1 And (DirectCast(sender, Button).Enabled) Then
            DirectCast(sender, Button).Enabled = False
            DirectCast(sender, Button).BackColor = Color.Black
            blackpoint.Add(1)
        End If

        If turn = 0 And (DirectCast(sender, Button).Enabled) Then
            DirectCast(sender, Button).Enabled = False
            DirectCast(sender, Button).BackColor = Color.White
            whitepoint.Add(1)
        End If



        colorchange(j)
    End Sub

    Private Sub Button6_Click(sender As Object, e As EventArgs) Handles Button6.Click
        Button1.Enabled = True
        Button2.Enabled = True
        Button3.Enabled = True
        Button4.Enabled = True

        Button1.BackColor = Color.Green
        Button2.BackColor = Color.Green
        Button3.BackColor = Color.Green
        Button4.BackColor = Color.Green

        whitepoint.Clear()
        blackpoint.Clear()
        TextBox1.Text = ""

        i = 0
        Label1.Text = Int(i)

    End Sub

    Private Sub Button7_Click(sender As Object, e As EventArgs) Handles Button7.Click
        whitesum = whitepoint.Sum
        blacksum = blackpoint.Sum

        If whitesum = blacksum Then
            TextBox1.Text = "引き分け"
        End If

        If whitesum > blacksum Then
            TextBox1.Text = "白の勝ち"
        End If

        If whitesum < blacksum Then
            TextBox1.Text = "黒の勝ち"
        End If
    End Sub
    
    ''' <summary>
    ''' fushiana 修正版
    ''' </summary>
    'Private i As Integer
    'Dim buttons As List(Of Button)

    'Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load
    '    buttons = New List(Of Button) From {
    '        Button1,
    '        Button2,
    '        Button3,
    '        Button4
    '    }
    'End Sub

    'Private Sub button_click(ByVal sender As Object, e As EventArgs) Handles Button1.Click, Button2.Click, Button3.Click, Button4.Click
    '    DirectCast(sender, Button).Enabled = False

    '    i += 1
    '    Label1.Text = i.ToString()
    '    Dim clickedButton = DirectCast(sender, Button)

    '    If (i Mod 2) = 1 And clickedButton.Enabled Then
    '        clickedButton.BackColor = Color.Black
    '    Else
    '        clickedButton.BackColor = Color.White
    '    End If
    '    ButtonColorCcheck(clickedButton)
    'End Sub


    'Private Sub ButtonColorCcheck(clickedButton As Button)
    '    Dim clickedButtons = buttons.Where(Function(x) x.Enabled = False And x.BackColor = clickedButton.BackColor).ToList()

    '    Dim firstButton = buttons.FindIndex(Function(x) x.BackColor = clickedButton.BackColor)
    '    Dim latestButton = buttons.FindLastIndex(Function(x) x.BackColor = clickedButton.BackColor)

    '    ButtonColorChenge(firstButton, latestButton, clickedButton.BackColor)
    'End Sub


    'Private Sub ButtonColorChenge(fromPosition As Integer, toPosition As Integer, backColor As Color)
    '    Dim targetButtons = buttons.GetRange(fromPosition, toPosition - fromPosition)
    '    For Each item In targetButtons
    '        item.BackColor = backColor
    '    Next
    'End Sub

    'Private Sub Button6_Click(sender As Object, e As EventArgs) Handles Button6.Click
    '    For Each item In buttons
    '        item.Enabled = True
    '        item.BackColor = Color.Green
    '    Next

    '    i = 0

    '    TextBox1.Text = String.Empty
    '    Label1.Text = i.ToString()
    'End Sub

    'Private Sub Button7_Click(sender As Object, e As EventArgs) Handles Button7.Click
    '    Dim whitesum = buttons.Where(Function(x) x.BackColor = Color.White).ToList.Count
    '    Dim blacksum = buttons.Where(Function(x) x.BackColor = Color.Black).ToList.Count

    '    If whitesum = blacksum Then
    '        TextBox1.Text = "引き分け"
    '        Return
    '    End If

    '    If whitesum > blacksum Then
    '        TextBox1.Text = "白の勝ち"
    '        Return
    '    End If

    '    If whitesum < blacksum Then
    '        TextBox1.Text = "黒の勝ち"
    '        Return
    '    End If
    'End Sub
End Class

File: Form1.cs
using System;
using System.Collections.Generic;
using System.Collections;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace miniosero_c
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        public int i;
        List<int> button_push = new List<int>();
        int button_sum;
        List<Color> btname = new List<Color>();
        List<int> blackpoint = new List<int>();
        List<int> whitepoint = new List<int>();


        private void Form1_Load(object sender, EventArgs e)
        {
            btname.Add(button1.BackColor);
            btname.Add(button2.BackColor);
            btname.Add(button3.BackColor);
            btname.Add(button4.BackColor);
            btname.Add(button5.BackColor);

            for (int a = 0; a <= 4; a++)
            {
                btname[a] = Color.Green;
            }

        }

        private void Button_clicked(object sender, EventArgs e)
        {
            Button btn = sender as Button;

            i = i + 1;

            int turn = i % 2;

            if (btn.Name.Equals("button1") || btn.Name.Equals("button2") || btn.Name.Equals("button3") || btn.Name.Equals("button4") || btn.Name.Equals("button5"))
            {
                {
                    btn.Enabled = false;
                }

                if (turn == 1)
                {
                    btn.BackColor = Color.Black;
                    blackpoint.Add(1);
                }
                else
                {
                    btn.BackColor = Color.White;
                    whitepoint.Add(1);
                }
            }

            button_push.Add(1);

            label2.Text = Convert.ToString(i);

            button_sum = button_push.Sum();


            if (button_sum <= 2)

            {
                textBox1.Text = "ボタンを押してください";
            }


            else if ((button_sum == 3) && (button1.BackColor == button3.BackColor))
            {
                button2.BackColor = btn.BackColor;
            }
            else if ((button_sum == 3) && (button2.BackColor == button4.BackColor))
            {
                button3.BackColor = btn.BackColor;
            }
            else if ((button_sum == 3) && (button3.BackColor == button5.BackColor))
            {
                button4.BackColor = btn.BackColor;
            }
            else if ((button_sum == 4) && (button1.BackColor == button4.BackColor))
            {
                button2.BackColor = btn.BackColor;
                button2.BackColor = button3.BackColor;
            }
            else if ((button_sum == 4) && (button2.BackColor == button5.BackColor))
            {
                button3.BackColor = btn.BackColor;
                button3.BackColor = button4.BackColor;
            }
            else if ((button_sum == 4) && (button1.BackColor == button3.BackColor))
            {
                button2.BackColor = btn.BackColor;
            }
            else if ((button_sum == 4) && (button2.BackColor == button4.BackColor))
            {
                button3.BackColor = btn.BackColor;
               
            }
            else if ((button_sum == 4) && (button3.BackColor == button5.BackColor))
            {
                button4.BackColor = btn.BackColor;
            }
            else if ((button_sum == 5) && (button1.BackColor == button5.BackColor))
            {
                for (int b = 0; b <= 4; b++)
                {
                    btname[b] = btn.BackColor;
                }
                textBox1.Text = "true";
            }
            else if ((button_sum == 5) && (button1.BackColor == button4.BackColor))
            {
                for (int c = 0; c <= 3; c++)
                {
                    btname[c] = btn.BackColor;
                }
            }
            else if ((button_sum == 5) && (button2.BackColor == button5.BackColor))
            {
                for (int d = 1; d <= 4; d++)
                {
                    btname[d] = btn.BackColor;
                }
                textBox1.Text = "true2";
         
            }
            else if ((button_sum == 5) && (button1.BackColor == button3.BackColor))
            {
                button2.BackColor = btn.BackColor;
                textBox1.Text = "true3";
  
            }
            else if ((button_sum == 5) && (button2.BackColor == button4.BackColor))
            {
                button3.BackColor = btn.BackColor;
                textBox1.Text = "true4";
            }
            else if ((button_sum == 5) && (button3.BackColor == button5.BackColor))
            {
                button4.BackColor = btn.BackColor;
                textBox1.Text = "true5";

            }
            else
            {

            }
                    
            label3.Text = Convert.ToString(button_sum);

        }  

        private void button6_Click(object sender, EventArgs e)
        {
            Application.Restart();
        }

        private void button7_Click(object sender, EventArgs e)
        {
            i = i + 1;
            label2.Text = Convert.ToString(i);
        }

        private void button8_Click(object sender, EventArgs e)
        {
            int blackpoint_sum;
            int whitepoint_sum;
            if (button_sum <= 2) 

            { 
                blackpoint_sum = blackpoint.Sum();
                whitepoint_sum = whitepoint.Sum();
                label7.Text = Convert.ToString(blackpoint_sum);
                label8.Text = Convert.ToString(whitepoint_sum);
            }

            else if (button_sum >= 3)
            {
                blackpoint_sum = (int)blackpoint.Last();
                whitepoint_sum = (int)whitepoint.Last();
                label7.Text = Convert.ToString(blackpoint_sum);
                label8.Text = Convert.ToString(whitepoint_sum);
            }
        }
                  
        
    }
}
